packages/remix-dev/__tests__/routesConvention-test.ts

```diff
@@ -36,6 +36,47 @@ describe("createRoutePath", () => {
       ["[index]", "index"],
       ["test/inde[x]", "test/index"],
       ["[i]ndex/[[].[[]]", "index/[/[]"],
+
+      // Optional segment routes
+      ["(routes)/$", "routes?/*"],
+      ["(routes)/($)", "routes?/*?"], // TODO: Fails, do we want to allow this?
+      ["(routes)/(sub)/$", "routes?/sub?/*"],
+      ["(routes).(sub)/$", "routes?/sub?/*"],
+      ["(routes.sub)/$", "routes?/sub?/*"], // TODO: Fails, do we want to allow this?
+      ["(routes)/($slug)", "routes?/:slug?"],
+      ["(routes)/sub/($slug)", "routes?/sub/:slug?"],
+      ["(routes).sub/($slug)", "routes?/sub/:slug?"],
+      ["($)", "*?"], // TODO: Fails, do we want to allow this?
+      ["(nested)/$", "nested?/*"],
+      ["(flat).$", "flat?/*"],
+      ["($slug)", ":slug?"],
+      ["(nested)/($slug)", "nested?/:slug?"],
+      ["(flat).($slug)", "flat?/:slug?"],
+      ["flat.(sub)", "flat/sub?"],
+      ["(nested)/(index)", "nested?"], // Fails with `"flat?/index?"`
+      ["(flat).(index)", "flat?"], // Fails with `"flat?/index?"`
+      ["(index)", undefined], // Fails with `"index?"`"
+      ["(__layout)/(index)", undefined], // Fails with __layout?/index?
+      ["__layout/(test)", "test?"],
+      ["__layout.(test)", "test?"],
+      ["__layout/($slug)", ":slug?"],
+      ["(nested)/__layout/($slug)", "nested?/:slug?"],
+      ["($slug[.]json)", ":slug.json?"],
+      ["(sub)/([sitemap.xml])", "sub?/sitemap.xml?"],
+      ["(sub)/[(sitemap.xml)]", "sub?/sitemap.xml?"], // TODO should this have been sub?/(sitemap.xml)
```

Comment:
yeah I think this looks like a bug since the `[]` escaping isn't working right

Reply:
Working now!